Library/OcAudioLib/OcAudioLib.c

```diff
@@ -89,19 +100,31 @@ OcAudioInstallProtocols (
       }
     }
   } else {
+    OcAudioProtocol     = NULL;
+    FoundOtherProtocols = FALSE;
     for (Index = 0; Index < ARRAY_SIZE (mAudioProtocols); ++Index) {
       Status = gBS->LocateProtocol (
         mAudioProtocols[Index],
         NULL,
         &Protocol
         );
       if (!EFI_ERROR (Status)) {
+        //
+        // Don't exit early, so we can log all found protocols.
```

Comment:
Guard with DEBUG_CODE_BEGIN? No need to log it in RELEASE builds.

Reply:
I think I really want to `break;` at the end of `!EFI_ERROR (Status) { ... }` block, but only on RELEASE build. Is there something like RELEASE_CODE_BEGIN?

Reply:
Or, as I think there is not, is it okay to add it in OC, modelled on DEBUG_CODE_BEGIN but with `!DebugCodeEnabled ()`??

Reply:
I think what you might want here is to log protocols in the DEBUG build and return the right protocol in both builds. Perhaps split the loops, leaving the logging one outside in DEBUG_CODE brackets? This will make the code cleaner.

Reply:
Is this okay: https://github.com/acidanthera/OpenCorePkg/blob/1d60f1d07a6a2116d7708b0e715dc34b7328e263/Library/OcAudioLib/OcAudioLib.c#L116-L121 ?

Reply:
Well, I personally would prefer my solution. It would be shorter and cleaner for RELEASE builds. Note, the code requires reordering OC protocol as the last one:

```C
    DEBUG_CODE_BEGIN ();
    for (Index = 0; Index < ARRAY_SIZE (mAudioProtocols); ++Index) {
      Status = gBS->LocateProtocol (
        mAudioProtocols[Index],
        NULL,
        &Protocol
        );
      DEBUG ((DEBUG_INFO, "OCAU: %g protocol - %r\n", mAudioProtocols[Index], Status));
    }
    DEBUG_CODE_END ();
    for (Index = 0; Index < ARRAY_SIZE (mAudioProtocols) - 1; ++Index) {
      Status = gBS->LocateProtocol (
        mAudioProtocols[Index],
        NULL,
        &Protocol
        );
      if (!EFI_ERROR (Status) {
        return NULL;
      }
    }
    Status = gBS->LocateProtocol (
      mAudioProtocols[ARRAY_SIZE (mAudioProtocols) - 1],
      NULL,
      &Protocol
      );
    if (!EFI_ERROR (Status)) {
      return Protocol;
    }
```

Reply:
The meaning changed in the above (should return OC protocol if found, else null if anything else found, else go ahead and install OC protcol). I've gone for this and not reordering the list, I think it is fine:

```C
    DEBUG_CODE_BEGIN ();
    for (Index = 0; Index < ARRAY_SIZE (mAudioProtocols); ++Index) {
      Status = gBS->LocateProtocol (
        mAudioProtocols[Index],
        NULL,
        &Protocol
        );
      DEBUG ((DEBUG_INFO, "OCAU: %g protocol - %r\n", mAudioProtocols[Index], Status));
    }
    DEBUG_CODE_END ();
    for (Index = 0; Index < ARRAY_SIZE (mAudioProtocols); ++Index) {
      Status = gBS->LocateProtocol (
        mAudioProtocols[Index],
        NULL,
        &Protocol
        );
      if (!EFI_ERROR (Status)) {
        if (Index == 0) {
          return (OC_AUDIO_PROTOCOL *) Protocol;
        }
        return NULL;
      }
    }
```